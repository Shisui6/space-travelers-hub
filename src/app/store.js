import { configureStore } from '@reduxjs/toolkit';
import rocketsReducer from '../features/RocketList/rocketsSlice';

const store = configureStore({
  reducer: {
    rockets: rocketsReducer,
  },
});

export default store;
