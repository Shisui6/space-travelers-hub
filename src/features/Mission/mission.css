.btn {
  background-color: #025fac;
  border: none;
  outline: none;
  color: #fff;
  border-radius: 6px;
  font-size: 15px;
  padding: 10px;
}

.badge {
  background-color: #177fd4;
  color: #fff;
  padding: 4px;
  margin-right: 3px;
  border-radius: 4px;
  font-size: 12px;
}

.cancel-btn {
  background-color: #fff;
  border: 1px solid rgba(51, 51, 51, 0.9);
  color: rgba(51, 51, 51, 0.9);
  border-radius: 6px;
  font-size: 15px;
  padding: 10px;
}

/* profile */
.profile {
  display: flex;
  justify-content: space-between;
  width: 100%;
}

.profile-section {
  flex: 1;
  margin: 18px;
}

.profile-list {
  width: 100%;
}

.profile-title {
  margin: 8px 0;
}

.list-item {
  border: 1px solid #bebfc5;
  padding: 8px;
}

/* mission */
.section {
  padding: 0 50px 50px;
}

.width-10 {
  width: 10%;
}

.width-15 {
  width: 15%;
  text-align: center;
}

.width-60 {
  width: 60%;
}

.mission-table {
  width: 100%;
  border: 1px solid #dee2e6;
}

.mission-row {
  display: flex;
  margin: 18px 8px;
}

.centerAlign {
  text-align: center;
}

.description {
  padding: 20px;
  text-align: justify;
}

tr:nth-child(even) {
  background: #fff;
}

tr:nth-child(odd) {
  background: #f2f2f2;
}

.badge-join {
  background-color: #6d757d;
  color: #fff;
  padding: 5px;
  margin-right: 3px;
  border-radius: 4px;
  font-size: 12px;
}

.badge-leave {
  background-color: #18a2b8;
  color: #fff;
  padding: 5px;
  margin-right: 3px;
  border-radius: 4px;
  font-size: 12px;
}

.btn-join {
  background-color: #fff;
  border: 1px solid rgba(51, 51, 51, 0.9);
  color: rgba(51, 51, 51, 0.9);
  border-radius: 5px;
  font-size: 15px;
  padding: 8px;
  cursor: pointer;
}

.btn-leave {
  background-color: #fff;
  border: 1px rgb(238, 5, 5) solid;
  outline: none;
  color: rgb(238, 5, 5);
  border-radius: 5px;
  font-size: 15px;
  padding: 8px;
  cursor: pointer;
}
