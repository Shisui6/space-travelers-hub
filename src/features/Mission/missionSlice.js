/* eslint-disable no-param-reassign */
import { createAsyncThunk, createSlice } from '@reduxjs/toolkit';

const GET_MISSIONS = 'missions/getMissions';

const API_URL = 'https://api.spacexdata.com/v3/missions';

const fetchMissions = async () => {
  const response = await fetch(API_URL);
  const missions = await response.json();

  return missions.map((obj) => ({
    id: obj.mission_id,
    name: obj.mission_name,
    description: obj.description,
    reserved: false,
  }));
};

export const getMissions = createAsyncThunk(GET_MISSIONS,
  async () => {
    const response = await fetchMissions();
    return response;
  });

const slice = createSlice({
  name: 'missions',
  initialState: {
    missions: [],
  },
  reducers: {
    joinMission(state, action) {
      const id = state.missions.findIndex(((mission) => (mission.id === action.payload.id)));
      state.missions[id].reserved = true;
    },
    leaveMission(state, action) {
      const id = state.missions.findIndex(((mission) => (mission.id === action.payload.id)));
      state.missions[id].reserved = false;
    },
  },
  extraReducers: (builder) => {
    builder.addCase(getMissions.fulfilled, (state, action) => {
      state.missions = action.payload;
    });
  },
});

export const { joinMission, leaveMission } = slice.actions;
export default slice.reducer;
