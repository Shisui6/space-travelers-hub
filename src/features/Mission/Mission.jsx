/* eslint-disable object-curly-newline */
/* eslint-disable comma-dangle */
import React from 'react';
import PropTypes from 'prop-types';
import { useDispatch } from 'react-redux';
import { joinMission, leaveMission } from './missionSlice';
import './mission.css';

const Mission = ({ mission }) => {
  const { id, name, description, reserved } = mission;

  const dispatch = useDispatch();

  return (
    <tr>
      <td className="centerAlign">{name}</td>
      <td className="description">{description}</td>
      <td className="centerAlign">
        <span>
          {reserved ? (
            <span className="badge-leave">ACTIVE MEMBER</span>
          ) : (
            <span className="badge-join">NOT A MEMBER</span>
          )}
        </span>
      </td>
      <td className="centerAlign">
        {reserved
          ? <button type="submit" className="btn-leave" onClick={() => dispatch(leaveMission({ id }))}>Leave Mission</button>
          : <button type="submit" className="btn-join" onClick={() => dispatch(joinMission({ id }))}>Join Mission</button>}
      </td>
    </tr>
  );
};

Mission.propTypes = {
  mission: PropTypes.shape({
    id: PropTypes.string,
    name: PropTypes.string,
    description: PropTypes.string,
    reserved: PropTypes.bool,
  }),
};

Mission.defaultProps = {
  mission: null,
};

export default Mission;
