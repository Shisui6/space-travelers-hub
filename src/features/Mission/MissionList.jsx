import React, { useEffect } from 'react';
import { useSelector, useDispatch } from 'react-redux';
import Mission from './Mission';
import { getMissions } from './missionSlice';
import './mission.css';

const MissionList = () => {
  const missions = useSelector((state) => state.missions);
  const dispatch = useDispatch();
  useEffect(() => {
    dispatch(getMissions());
  }, [dispatch]);

  return (
    <section className="section">
      <table className="mission-table">
        <thead>
          <th className="width-10">Mission</th>
          <th className="width-60">Description</th>
          <th className="width-15">Status</th>
          <th className="width-15">Action</th>
        </thead>
        <tbody>
          {missions.map((mission) => (
            <Mission key={mission.id} mission={mission} />
          ))}
        </tbody>
      </table>
    </section>
  );
};

export default MissionList;
