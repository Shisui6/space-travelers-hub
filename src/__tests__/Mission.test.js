import React from 'react';
import { Provider } from 'react-redux';
import { render } from '@testing-library/react';
import store from '../app/store';
import Mission from '../features/Mission/Mission';

test('Testing Mission with snapshot', async () => {
  const data = {
    mission_id: 2,
    mission_name: 'Telstar',
    description:
      'Telstar 19V (Telstar 19 Vantage) is a communication satellite in the Telstar series of the Canadian satellite communications company Telesat. It was built by Space Systems Loral (MAXAR) and is based on the SSL-1300 bus. As of 26 July 2018, Telstar 19V is the heaviest commercial communications satellite ever launched, weighing at 7,076 kg (15,600 lbs) and surpassing the previous record, set by TerreStar-1 (6,910 kg/15230lbs), launched by Ariane 5ECA on 1 July 2009.',
    reserved: false,
  };
  const tree = render(
    <Provider store={store}>
      <Mission mission={data} />
    </Provider>,
  );

  expect(tree).toMatchSnapshot();
});
