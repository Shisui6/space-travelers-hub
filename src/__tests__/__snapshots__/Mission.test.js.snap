// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Testing Mission with snapshot 1`] = `
<tr>
  <td
    className="centerAlign"
  />
  <td
    className="description"
  >
    Telstar 19V (Telstar 19 Vantage) is a communication satellite in the Telstar series of the Canadian satellite communications company Telesat. It was built by Space Systems Loral (MAXAR) and is based on the SSL-1300 bus. As of 26 July 2018, Telstar 19V is the heaviest commercial communications satellite ever launched, weighing at 7,076 kg (15,600 lbs) and surpassing the previous record, set by TerreStar-1 (6,910 kg/15230lbs), launched by Ariane 5ECA on 1 July 2009.
  </td>
  <td
    className="centerAlign"
  >
    <span>
      <span
        className="badge-join"
      >
        NOT A MEMBER
      </span>
    </span>
  </td>
  <td
    className="centerAlign"
  >
    <button
      className="btn-join"
      onClick={[Function]}
      type="submit"
    >
      Join Mission
    </button>
  </td>
</tr>
`;
